docs/development/core/development-forking.asciidoc

```diff
@@ -0,0 +1,57 @@
+
+**Scenario 2: The thing we want mostly exists in viable third-party projects, but they don't do exactly what we want.**
+
+Consider filling in the gaps in functionality in these third party projects.
+
+Maybe the alternative project would accept contributions for the changes we desire. This would be the best scenario.
+
+Alternatively, perhaps there's a lightweight module we could develop in Kibana that uses this third-party module as an implementation detail so we don't have to build it all from scratch, but we can add the additional capabilities we need in our own code. The benefit of this approach is that we could move forward with the changes in Kibana now while still working to contribute the changes upstream so our custom module could potentially be short lived.
+
+**Scenario 3: We require a change in an existing third-party dependency, but that dependency has since been abandoned or is refusing to accept the change we need.**
+
+Ouch. This is the nightmare scenario we constantly try to avoid because there isn't an easy answer.
+
+First, accept that there is *not* a quick solution here.
+
+Next, accept that the only sustainable, long-term solution here is to move away from this third party dependency one way or another.  To that end, audit the usage of this dependency in kibana and explore alternative third-party dependencies to migrate to.  If there's a plausible alternative, then see scenario 2.
+
+It's rare that a drop-in replacement will exist for any dependency, so migrating from one dependency to another will require one of two approaches:
+
+1. Build a custom wrapper for the new module that we can use to shim the api of the abandoned module to the new module so that we do have a drop-in replacement as far as consuming code in Kibana is concerned. If you can repurpose the tests from the abandoned module to verify compatible APIs in your shim, then do so.
+2. Update all code that consumes this abandoned module in Kibana to use the new dependency. Make sure there's reasonable test coverage on the various features of Kibana that you are touching in order to prevent regressions when you do make the switch.
+
+It's rare that you'll encounter this scenario at a time when it is convenient to undergo this level of effort, but that doesn't mean we shouldn't undergo the effort. At this point, we have a ticking time bomb in Kibana that needs to be addressed now.
+
+[float]
+==== The cardinal forking sin
+
+Forking a dependency is almost certainly going to be rejected if the intention is to maintain a new repo with some customizations while continuing to sync changes with the original project over time (or ever again, really).
+
+[float]
+==== It's forking time
+
+If you absolutely *must* fork a project, then understand that the forked code is now our code, and we are responsible for supporting it as extensively as any that we wrote ourselves. In concrete terms, the forked source code...
+
+* should be brought into the kibana repo
```

Comment:
Instructions for doing this would be a huge plus. Are we going down the monorepo path? Is there a specific place where we should be dropping the fork?

Also, rules for how to create the fork would be helpful. Simply saying "copy module into the Kibana repo" is bad, because then it becomes impossible to track what changes were made in our fork...

Reply:
Good call on specifics here. My intention with this whole section was to emphasize that you are explicitly and forever breaking off from the original code and this is now kibana source code in the kibana repo, so there is no notion of tracking changes that we make versus the original, but based on your feedback I don't think that is at all clear.

Reply:
Yeah, I get the mindset. I'd say put exactly that sentence in there somewhere.